os/kernel/preference/preference_write.c

```diff
@@ -122,26 +123,26 @@ static int preference_write_fs_key(char *path, preference_data_t *data)
 		return PREFERENCE_IO_ERROR;
 	}
 
-	if (data->value.type != PREFERENCE_TYPE_STRING) {
-		/* Write type, len, value */
-		write_size = 2 * sizeof(int) + data->value.len;
-		ret = write(fd, (void *)&(data->value), write_size);
-	} else {
-		/* Write type, len */
-		write_size = 2 * sizeof(int);
-		ret = write(fd, (void *)&(data->value), write_size);
-		if (ret == write_size) {
-			/* Write value */
-			write_size += data->value.len;
-			ret += write(fd, (void *)data->value.s, data->value.len);
-		}
+	/* Calculate checksum of attributes, type, len and value */
+	crc_value = crc32((uint8_t *)&data->attr.type, sizeof(value_attr_t) - sizeof(uint32_t));
+	data->attr.crc = crc32part((uint8_t *)data->value, data->attr.len, crc_value);
+
+	/* Write attributes of data : crc, type, len */
+	ret = write(fd, (void *)&data->attr, sizeof(value_attr_t));
+	if (ret != sizeof(value_attr_t)) {
+		prefdbg("Failed to write key value, errno %d\n", errno);
+		goto errout_with_close;
 	}
-	if (ret != write_size) {
-		ret = PREFERENCE_IO_ERROR;
+
+	/* Write value data */
+	ret = write(fd, (void *)data->value, data->attr.len);
```

Comment:
You already checked validation and everything calculated.
Then you can write all just once. Internally, calling write twice is take bunch of processes.

Reply:
I called write function twice to avoid allocating buffer as a size of whole data and copying them because actual value is in another address, not continuous address region with another data.